sys/netbsd/lwp.txt

```diff
@@ -31,16 +31,16 @@ _lwp_ctl(features int32, address ptr[in, ptr[in, lwpctl]])
 
 resource semid[intptr]
 
-_ksem_init(val int32, idp ptr[out, semid])
-_ksem_open(name buffer[in], oflag flags[semopen_flags], mode flags[open_mode], val int32, idp ptr[in, semid])
+_ksem_init(val int32, idp ptr[inout, semid])
```

Comment:
Can we specify somehow that on the `in` entry there are only two valid numbers 0 and `KSEM_PSHARED`?

CC: @dvyukov

Reply:
The flags type can be used for this:
https://github.com/google/syzkaller/blob/master/docs/syscall_descriptions.md#flagsenums

Reply:
And even if the output value will be out of the flag entry, it will be fine?

@ais2397 please use enum-type flag here.

Reply:
Oh, you mean for the idp? Is it 0/KSEM_PSHARED on entry and the semid resource as output?
We don't have support for different types for input/output... I guess somebody wanted to save 4 bytes of memory 40 years ago and now we need to struggle with it :)

0 is a the default special value for any resource, we can explicitly state KSEM_PSHARED as well:
```
resource semid[intptr]: 0, KSEM_PSHARED
```

then we can declare the semid as optional here to increase chances of passing the special values (good in either case because we don't really need to pass an existing semid as input here):
```
idp ptr[inout, semid[opt]]
```

But then the fuzzer will pass KSEM_PSHARED for all other uses of semid from time to time.

Reply:
What is the value for KSEM_PSHARED? Maybe it's possible to play some bitfield tricks if semid's are usually small KSEM_PSHARED does not occupy low bits...

Reply:
You may leave a `NEED:` comment describing what is the feature we are failing to describe properly (grep for other examples).

Reply:
> Oh, you mean for the idp? Is it 0/KSEM_PSHARED on entry and the semid resource as output?

Yes.

On entry it is either 0 or KSEM_PSHARED, on output it is an unique intptr_t value.

> But then the fuzzer will pass KSEM_PSHARED for all other uses of semid from time to time.

As long as it will pass a valid semid for other entry calls at least in half of the cases, it should be enough.

Reply:
> What is the value for KSEM_PSHARED? Maybe it's possible to play some bitfield tricks if semid's are usually small and KSEM_PSHARED does not occupy low bits...

`#define	KSEM_PSHARED		0x50535244U	/* 'PSRD' */`

https://nxr.netbsd.org/xref/src/sys/sys/ksem.h#66

The returned semid value as far as I can see, is a random number between the range 0x0 and 0x01000000U with an optional bit set inside the 0xff000001U mask.

Reference:

https://nxr.netbsd.org/xref/src/sys/kern/uipc_sem.c

Reply:
0x50535244U probably won't allows us to describe some kind of bitfield. But:

> As long as it will pass a valid semid for other entry calls at least in half of the cases, it should be enough.

It will pass valid semid's even more frequently. These special values are for things like fd -1, which are not all that interesting to pass.